Unbreak various tests in CI

Zlopez opened the pull request:
This should fix the failing CI tests

Signed-off-by: Michal Konečný <mkonecny@redhat.com>

Zlopez committed:
Fix skopeo tests

Prior to pytest 5.0, ExceptionInfo objects had a __str__() method of
their own which rendered the wrapped exception as a string. It was
removed in version 5.0 which made calling str() on the object use
__repr__() as a fallback which gives totally different results.

Here we use str() on the .value member (i.e. the wrapped exception) and
a variable name that makes it (more) explicit that the context manager
isn't an exception.

Signed-off-by: Michal Konečný <mkonecny@redhat.com>
Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp commented:
I've updated the PR:

- Added a description to the commit log about what went wrong here, i.e. backwards-incompatible changes in pytest 5.0 triggered the test failures. Because 5.0 dropped support for Python 2.x, we don't have it in Fedora yet, so the problem only showed up in some setups (Python modules installed with pip and up-to-date).
- Renamed the context manager object from `e` to `excinfo` to make it more explicit we're not dealing with an exception here.
- Rather than `str(e.value), used the `ExceptionInfo.exconly()` method which exposes name and message of the wrapped exception (and do this consistently in another test with the second commit).

pypingou commented:
Ok, I'm confused the test results show as red here but visiting: https://ci.centos.org/job/bodhi-pipeline/job/PR-3371/3/ I see ``Test Result (no failures)``

is it still running?

nphilipp commented:
> Ok, I'm confused the test results show as red here but visiting: https://ci.centos.org/job/bodhi-pipeline/job/PR-3371/3/ I see `Test Result (no failures)`
> 
> is it still running?

Right now it still shows 6 as "expected". While the tests are in flight, I always look into the [Pipeline Steps](https://ci.centos.org/job/bodhi-pipeline/job/PR-3371/3/flowGraphTable) and, of the ones marked red, check their log (terminal symbol beside the red bubble). Basically, we have one issue in the pip-integration/-build tests with an unsigned package, and the sphinx problem in rawhide-docs.

pypingou reviewed (approved):
nice work tracking this down! :)

nphilipp committed:
Work around Fedora Infra RPM repo key import issue

This is an obscure and hard to debug (therefore unfixed) problem in DNF
which caused installing Fedora Infrastructure RPM packages to fail in CI
containters. It's documented at:

https://bugzilla.redhat.com/show_bug.cgi?id=1699396

Thanks to Brian Stinson for the suggested workaround!

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp commented:
With #3373 in flight I have had some time to think about `str(excinfo.value)` vs. `excinfo.exconly()`, especially that the latter isn't the same as the former but also includes the exception name. @Zlopez do you have an opinion?

Zlopez commented:
@nphilipp Will `excinfo.exconly()` work with current tests?

pypingou commented:
> Will excinfo.exconly() work with current tests?

Yes, we checked that, it was added in 2007 (iirc)

nphilipp commented:
As discussed, let's broaden the scope of this PR and include:

* Work around Fedora Infra RPM repo key import issue (#3373)
* Add `skip_if_unavailable` to the infra-tags repo on Rawhide (due to https://fedoraproject.org/wiki/Changes/Set_skip_if_unavailable_default_to_false from F31 on, this only triggered now)
* Clamp `pydocstyle < 4.0` because https://github.com/PyCQA/pydocstyle/issues/370 trips over the nested function in bodhi/server/renderers.py

nphilipp committed:
cope with DNF changes in Fedora >= 31

In previous versions, skip_if_unavailable defaulted to True so the
missing infrastructure-tags repository for the Rawhide release version
didn't make tests fail. Restore this setting for the infrastructure-tags
repository.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp commented:
Oh, and as discussed I've replaced `excinfo.exconly()` with `str(excinfo.value)` as it's closer to what was done originally.

pypingou commented on devel/ci/integration/bodhi/Dockerfile-rawhide:
```diff
@@ -7,6 +7,12 @@ LABEL \
+RUN rpm --import "$(awk -F= '{if ($0 ~ "gpgkey *=") {print $2;}}' < /etc/yum.repos.d/infra-tags.repo)"
+# deal with https://fedoraproject.org/wiki/Changes/Set_skip_if_unavailable_default_to_false
+RUN dnf install -y dnf-plugins-core
+RUN dnf config-manager infrastructure-tags --setopt skip_if_unavailable=True --save
```
Every command adds a layer in the container, do you know if Randy and Aurélien preferred the one command / line approach? Otherwise, it is often recommended, to have multiple command per "RUN" using ``\`` to delimit them.

nphilipp replied:
I don't know their preference. In the Dockerfiles there are instances of one RUN line containing multiple commands (though not many), but also multiple consecutive RUN lines each containing a single command.

nphilipp committed:
Clamp pydocstyle < 4.0 in CI tests

This is because pydocstyle 4.0.0 can't cope with arguments described in
the docstrings of nested functions, as in bodhi/server/renderers.py:

https://github.com/PyCQA/pydocstyle/issues/370

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp commented:
For reference, here's an example of `pydocstyle` tripping: https://ci.centos.org/job/bodhi-pipeline/job/PR-3371/6/execution/node/329/log/

Zlopez commented:
So only thing failing now will be rawhide docs test? Or this is a new issue?

nphilipp commented:
> So only thing failing now will be rawhide docs test? Or this is a new issue?

No, it's that `cornice_sphinx` is incompatible with the latest `sphinx` version: https://github.com/Cornices/cornice.ext.sphinx/issues/22

Zlopez commented:
I see that there are plenty of integration tests failing, but this will be probably related to https://github.com/fedora-infra/bodhi/pull/3370

nphilipp commented:
As discussed: considering that we verified the changes in this PR to fix what they're supposed to do and that they aren't causing the remaining and new CI pipeline failures, I'll attempt to merge them manually, which should make fixing the outstanding issues easier as you don't have to dig through the errors this one is fixing.
